internal/auth/auth.go

```diff
@@ -9,79 +9,139 @@ import (
-	claim, err := getTelemetryClaim(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("%s: %w", errUnauthorized.Error(), err)
+var manufacturerPrivMap = map[privileges.Privilege]model.Privilege{
+	privileges.ManufacturerDeviceLastSeen: model.PrivilegeManufacturerDeviceLastSeen,
+}
+
+type PrivilegeValidator struct {
+	VehicleNFTAddress      common.Address
+	ManufacturerNFTAddress common.Address
+}
+
+type TokenValidator struct {
+	IdentitySvc identity.IdentityService
+}
+
+// VehicleNFTPrivCheck checks if the claim set in the context includes the correct address the required privileges for the VehicleNFT contract.
+func (pcv *PrivilegeValidator) VehicleNFTPrivCheck(ctx context.Context, _ any, next graphql.Resolver, privs []model.Privilege) (any, error) {
+	if err := pcv.checkPrivWithAddress(ctx, privs, pcv.VehicleNFTAddress); err != nil {
+		return nil, fmt.Errorf("unauthorized: %w", err)
 	}
-	fCtx := graphql.GetFieldContext(ctx)
-	if fCtx == nil {
-		return nil, fmt.Errorf("%w: no field context found", errUnauthorized)
+
+	return next(ctx)
+}
+
+// ManufacturerNFTPrivCheck checks if the claim set in the context includes the correct address the required privileges for the ManufacturerNFT contract.
+func (pcv *PrivilegeValidator) ManufacturerNFTPrivCheck(ctx context.Context, obj interface{}, next graphql.Resolver, privs []model.Privilege) (res interface{}, err error) {
+	if err := pcv.checkPrivWithAddress(ctx, privs, pcv.ManufacturerNFTAddress); err != nil {
+		return nil, fmt.Errorf("unauthorized: %w", err)
 	}
-	tokenID, ok := fCtx.Args["tokenId"].(int)
-	if !ok {
-		return nil, fmt.Errorf("%w: failed to get tokenID from args", errUnauthorized)
+
+	return next(ctx)
+}
+
+// VehicleTokenCheck checks if the vehicle tokenID in the context matches the tokenID in the claim.
+func (tv *TokenValidator) VehicleTokenCheck(ctx context.Context, _ any, next graphql.Resolver) (any, error) {
+	vehicleTokenID, err := getArg[int](ctx, "tokenId")
+	if err != nil {
+		return nil, fmt.Errorf("unauthorized: %w", err)
 	}
-	if strconv.Itoa(tokenID) != claim.TokenID {
-		return nil, fmt.Errorf("%w: tokenID mismatch", errUnauthorized)
+
+	if err := headerTokenMatchesQuery(ctx, func() (string, error) {
+		return strconv.Itoa(vehicleTokenID), nil
+	}); err != nil {
+		return nil, fmt.Errorf("unauthorized: %w", err)
 	}
+
 	return next(ctx)
 }
 
-type PrivilegeContractValidator struct {
-	VehicleNFTAddress common.Address
-	// ManufAddr         string
+func (tv *TokenValidator) ManufacturerTokenCheck(ctx context.Context, obj interface{}, next graphql.Resolver) (res interface{}, err error) {
+	// TODO(ae) allow user to search by addr, tokenID, or serial
+	aftermarketDeviceAddr, err := getArg[common.Address](ctx, "address")
+	if err != nil {
+		return nil, fmt.Errorf("unauthorized: %w", err)
+	}
+
+	if err := headerTokenMatchesQuery(ctx, func() (string, error) {
+		resp, err := tv.IdentitySvc.AftermarketDevice(ctx, &aftermarketDeviceAddr, nil, nil)
```

Comment:
What happens if we can't find the aftermarket device? Is the error message nice?

Reply:
Let's implement all three!